File: archive/python/array/3SumClosest.py
class Solution:
    # @return an integer
    def threeSumClosest(self, num, target):
        num.sort()
        low = 0
        high = len(num) - 1
        closest = abs(num[low] + num[high] + num[1] - target)
        result = num[low] + num[high] + num[1]
        while high - low > 1:
            curr = num[low] + num[high]
            for i in xrange(low + 1, high):
                temp = curr + num[i]
                diff = abs(target - temp)
                if diff == 0:
                    return target
                if diff < closest:
                    closest = diff
                    result = temp
            if curr < target:
                low += 1
            elif curr > target:
                high -= 1
            else:
                #pay attention here
                if abs(num[high-1]) > abs(num[low+1]):
                    high -= 1
                else:
                    low += 1
        return result

if __name__ == "__main__":
    solution = Solution()

    #print solution.threeSumClosest([-3, -2, -5, 3, -4], -1)
    print solution.threeSumClosest([-5, -4, -3, -2, 3], -1)

File: archive/cpp/array/3SumClosest.cpp
//
//  3SumClosest.cpp
//  c++
//
//  Created by sure on 1/25/15.
//  Copyright (c) 2015 Shuo Li. All rights reserved.
//

#include <stdio.h>
#include <stdlib.h>
#include <vector>
#include <iostream>


class Solution {
public:
    int threeSumClosest(std::vector<int> &num, int target) {
        std::sort(num.begin(), num.end());
        int low = 0;
        int high = int(num.size() - 1);
        int result = num[0] + num[high] + num[1];
        int closest = abs(num[0] + num[high] + num[1] - target);
        while (high - low > 1) {
            int curr = num[low] + num[high];
            for (int i = low + 1; i < high; i ++) {
                int temp = curr + num[i];
                int diff = abs(target-temp);
                if (diff == 0) {
                    return target;
                }
                if (diff < closest) {
                    closest = diff;
                    result = temp;
                }
            }
            if (curr < target) {
                low ++;
            } else if (curr > target) {
                high --;
            } else {
                if (abs(num[high]) > abs(num[low])) {
                    high --;
                } else {
                    low ++;
                }
            }
        }
        return result;
    }
};
